rust/operator-binary/src/connect/server.rs

```diff
@@ -0,0 +1,586 @@
+                        &SparkConnectRole::Server.to_string(),
+                        DUMMY_SPARK_CONNECT_GROUP_NAME,
+                    )
+                    .context(LabelBuildSnafu)?
+                    .into(),
+                ),
+                ..LabelSelector::default()
+            },
+            ..DeploymentSpec::default()
+        }),
+        ..Deployment::default()
+    })
+}
+
+#[allow(clippy::result_large_err)]
+pub fn build_service(
+    scs: &v1alpha1::SparkConnectServer,
+    app_version_label: &str,
+    service_cluster_ip: Option<String>,
+) -> Result<Service, Error> {
+    let (service_name, service_type, publish_not_ready_addresses) = match service_cluster_ip.clone()
+    {
+        Some(_) => (
+            // These are the properties of the headless driver service used for the internal
+            // communication with the executors as recommended by the Spark docs.
+            //
+            // The flag `publish_not_ready_addresses` *must* be `true` to allow for readiness
+            // probes. Without it, the driver runs into a deadlock beacuse the Pod cannot become
+            // "ready" until the Service is "ready" and vice versa.
+            object_name(&scs.name_any(), SparkConnectRole::Server),
+            "ClusterIP".to_string(),
+            Some(true),
+        ),
+        None => (
+            format!(
+                "{}-{}",
+                object_name(&scs.name_any(), SparkConnectRole::Server),
+                SparkConnectRole::Server
+            ),
+            scs.spec.cluster_config.listener_class.k8s_service_type(),
+            Some(false),
+        ),
+    };
+
+    let selector = Labels::role_selector(scs, APP_NAME, &SparkConnectRole::Server.to_string())
+        .context(LabelBuildSnafu)?
+        .into();
+
+    Ok(Service {
+        metadata: ObjectMetaBuilder::new()
+            .name_and_namespace(scs)
+            .name(service_name)
+            .ownerreference_from_resource(scs, None, Some(true))
+            .context(ObjectMissingMetadataForOwnerRefSnafu)?
+            .with_recommended_labels(common::labels(
+                scs,
+                app_version_label,
+                &SparkConnectRole::Server.to_string(),
+            ))
+            .context(MetadataBuildSnafu)?
+            .with_label(Label::try_from(("prometheus.io/scrape", "true")).context(LabelBuildSnafu)?)
+            .build(),
+        spec: Some(ServiceSpec {
+            type_: Some(service_type),
+            cluster_ip: service_cluster_ip,
+            ports: Some(vec![
+                ServicePort {
+                    name: Some(String::from("grpc")),
+                    port: CONNECT_GRPC_PORT,
+                    ..ServicePort::default()
+                },
+                ServicePort {
+                    name: Some(String::from("http")),
+                    port: CONNECT_UI_PORT,
+                    ..ServicePort::default()
+                },
+            ]),
```

Comment:
`grpc` and `http` are used as well in `build_deployment` as well, constant?

Reply:
https://github.com/stackabletech/spark-k8s-operator/pull/539/commits/a95673098dc2e3997b3a6adc24e6d11aa78883b4